### ISSUE TITLE
useReplicache

### ISSUE THREAD

**Issue by arv** [rocicorp/replicache#405]
Add a useReplicache to replicache-react-util that deals with close/open races

**Comment by arv** [comment]
My initial idea was to keep track of closing `Replicache` instances and await their close when you create a new one with the same name.

There is one small complication and that is that `useReplicache` and `new Replicache` is sync so we would need to make all existing async methods also await the closing/reopen promise. We can do this with a wrapper. The wrapper can maybe done using a `Proxy` or we could subclass. Something along the lines of:

```ts
class UseReplicache extends Replicache {
   #reopenPromise: Promise<unknown>;

   ...

  override async get(key: string) {
    await this.#reopenPromise;
    return super.get(key);
  }
}
```

This gets pretty brittle and is very verbose so maybe a better way is to extend the API to allow passing in a `Promise` to await before opening the database. Or make `_open` public/protected, which would allow overriding.

**Comment by aboodman** [comment]
Maybe stupid question, but do we need `close()`? A simpler solution would be to allow only one Replicache instance per name per JS context?

**Comment by aboodman** [comment]
> we would need to make all existing async methods also await the closing/reopen promise. 

We kind of already have this because various places in `Replicache` await `this._openResponse` or `this._root` (not sure if we need both?).

I like this pattern because TypeScript ensures that we cannot screw it up because you have to await the promise to get at the data you need.

**Comment by arv** [comment]
> do we need `close()`?

I think we do because we do not want the connection loops running if you "sign out". Another use case is to disable/pause syncing and that is today done by closing the Replicache instance (of course would provide an alternative API)

> maybe a better way is to extend the API

My thinking is to do something like:

```ts
export interface ReplicacheOptions<MD extends MutatorDefs> {
  ...
  openPromise?: Promise<unknown>;
}
```

and then inside `_open` do

```ts
  private async _open(): Promise<void> {
    await this._options.openPromise; // we do not really have an `_options` property...
    ...
  }
```

I think the only question is if it is worth polluting the API with this?

**Comment by aboodman** [comment]
> I think the only question is if it is worth polluting the API with this?

Is it not possible to make it package-internal somehow? If not, if it makes everything else cleaner I'm fine with it.

**Comment by aboodman** [comment]
Also couldn't we (theoretically) hide it in a global map keyed by instance name, rather than passing it through the constructor?

**Comment by arv** [comment]
> Also couldn't we (theoretically) hide it in a global map keyed by instance name, rather than passing it through the constructor?

That gave me an idea. I'll upload some PRs shortly

**Comment by arv** [comment]
I ended up with my original idea :-) But the thing that became clear was that it needs to be done inside replicache and not inside replicache-react-util.

**Comment by KeKs0r** [comment]
I am wondering if there are other scenarios where this is relevant except for hot reloading. 
And if that is the main use case, there could be a solution instead of open and closing the instance to reuse it.
 I am just not sure if this could be done elegantly within replicache or if it requires more complex configuration in userland.

My initial gut feeling is that a clean open and closing within the react lifecycle sounds cleaner, but maybe there is a simple and elegant way of skipping the close altogether.

**Comment by arv** [comment]
> instead of open and closing the instance to reuse it.

Application code can decide to use a `Map<string, Replicache>` and never close the instances.

Closing the instance has some uses: 
 - subscriptions are finalized using `onDone`
 - The connection loops are terminated (no more requests are sent)
 - Stop listening to changes from other tabs

These things are important if your application supports user switching or if the app needs to stop Replicache for whatever reason. With Vercel we stopped Replicache when the actual keynote started.
